File: frontend/src/engine/capabilities/createTable.ts
/**
 * createTable – Convert a range into a structured Excel Table.
 *
 * Office.js notes:
 * - Tables provide auto-filter, structured references, and styling.
 * - Table names must be unique in the workbook.
 * - hasHeaders indicates whether the first row is a header row.
 * - Table styles are predefined strings like "TableStyleMedium2".
 */

import { CapabilityMeta, CreateTableParams, StepResult, ExecutionOptions } from "../types";
import { registry } from "../capabilityRegistry";
import { resolveRange, resolveSheet } from "./rangeUtils";

const meta: CapabilityMeta = {
  action: "createTable",
  description: "Convert a range into a structured Excel Table",
  mutates: true,
  affectsFormatting: true,
  requiresApiSet: "ExcelApi 1.1",
};

async function handler(
  context: Excel.RequestContext,
  params: CreateTableParams,
  options: ExecutionOptions
): Promise<StepResult> {
  const { range: address, hasHeaders = true, style } = params;
  // Auto-generate a unique table name if not provided
  const tableName = params.tableName || `Table_${Date.now()}`;

  if (options.dryRun) {
    return {
      stepId: "",
      status: "success",
      message: `Would create table "${tableName}" from ${address}`,
    };
  }

  options.onProgress?.(`Creating table "${tableName}"...`);

  // Resolve sheet from the range address so this works on any sheet,
  // not just the currently active one.
  const range = resolveRange(context, address);
  const sheet = resolveSheet(context, address);
  const table = sheet.tables.add(range, hasHeaders);

  table.name = tableName;

  if (style) {
    table.style = style;
  }

  table.load(["name", "style"]);
  await context.sync();

  return {
    stepId: "",
    status: "success",
    message: `Created table "${table.name}" with style ${table.style}`,
    outputs: { tableName: table.name, tableRange: address },
  };
}


registry.register(meta, handler as any);
export { meta };

File: backend/app/execution/capabilities/create_table.py
"""createTable — convert a range to an Excel table (ListObject)."""

from __future__ import annotations

from typing import Any

from app.execution.base import ExecutorContext
from app.execution.capability_registry import registry
from app.execution.range_utils import resolve_range


def handler(ctx: ExecutorContext, params: dict[str, Any]) -> dict[str, Any]:
    address = params.get("range")
    name = params.get("tableName") or params.get("name")
    has_headers = bool(params.get("hasHeaders", True))
    style = params.get("tableStyle") or "TableStyleMedium2"
    if not address:
        return {"status": "error", "message": "createTable requires 'range'."}

    rng = resolve_range(ctx.workbook_handle, address)
    if ctx.dry_run:
        return {"status": "preview", "message": f"Would create table on {rng.address}."}

    try:
        # Sheet.api.ListObjects.Add(SourceType=xlSrcRange, Source=range, XlListObjectHasHeaders=xlYes)
        list_object = rng.sheet.api.ListObjects.Add(
            1,  # xlSrcRange
            rng.api,
            None,
            1 if has_headers else 2,  # xlYes | xlNo
        )
        if name:
            list_object.Name = name
        list_object.TableStyle = style
    except Exception as exc:  # noqa: BLE001
        return {"status": "error", "message": f"Create table failed: {exc}", "error": str(exc)}

    return {
        "status": "success",
        "message": f"Created table on {rng.address}.",
        "outputs": {"tableName": list_object.Name, "range": rng.address},
    }


registry.register("createTable", handler, mutates=True, affects_formatting=True)
